File: buzu/src/test/java/br/net/buzu/metadata/ComplexMetadataTest.java
package br.net.buzu.metadata;

import br.net.buzu.java.lang.Syntax;
import br.net.buzu.java.model.Domain;
import br.net.buzu.java.model.MetaInfo;
import br.net.buzu.java.model.Metadata;
import br.net.buzu.java.model.Subtype;
import org.junit.Test;

import java.util.ArrayList;
import java.util.List;

import static br.net.buzu.java.model.Domains.domainOf;
import static org.junit.Assert.*;

/**
 * @author Douglas Siviotti
 *
 */
public class ComplexMetadataTest {

	public static ComplexMetadata createSample(String name, Subtype subtype, int size, int scale, int minOccurs,
			int maxOccurs, Domain domain, String defaultValue, String tags, List<Metadata> children) {
		MetaInfo metaInfo = new MetaInfo(31, name, subtype, size, 0, minOccurs, maxOccurs, domain, defaultValue, tags);
		return new ComplexMetadata(metaInfo, children);

	}

	static ComplexMetadata createSample(String name, Subtype subtype, int size, int minOccurs, int maxOccurs) {
		Domain domain = domainOf("white", "black", "red");
		return createSample(name, subtype, size, 0, minOccurs, maxOccurs, domain, "red", "XYZ", null);

	}

	@Test
	public void test() {
		SimpleMetadata m1 = SimpleMetadataTest.createSample("atomic", Subtype.CHAR, 10, 0, 1);
		SimpleMetadata m2 = SimpleMetadataTest.createSample("array", Subtype.CHAR, 10, 1, 2);
		List<Metadata> children = new ArrayList<>();
		children.add(m1);
		children.add(m2);
		ComplexMetadata c1 = createSample("complex", Subtype.OBJ, 0, 0, 0, 1, Domain.Companion.getEMPTY(), Syntax.EMPTY, Syntax.EMPTY, children);

		assertEquals(m1, c1.children().get(0));
		assertEquals(m2, c1.children().get(1));
		//assertEquals(30, c1.serialMaxSize()); // m1 = 10 m2 = 20

		// Basic Methods
		ComplexMetadata c2 = createSample("complex", Subtype.OBJ, 0, 0, 0, 1, Domain.Companion.getEMPTY(), Syntax.EMPTY, Syntax.EMPTY, children);

		assertTrue(c1.equals(c2));
		assertTrue(c1.hashCode() == c2.hashCode());
		assertTrue(c1.toString().equals(c2.toString()));
		assertTrue(c2.equals(c1));
		assertTrue(c1.equals(c1));
		assertFalse(c1.equals(null));
		Object o = "abc";
		assertFalse(c1.equals(o));
	}

}

File: ppl-lib/src/test/kotlin/br/net/buzu/pplimpl/metadata/ComplexMetadataTest.kt
package br.net.buzu.pplimpl.metadata

import br.net.buzu.pplspec.lang.EMPTY
import br.net.buzu.pplspec.model.*
import java.util.*
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test



/**
 * @author Douglas Siviotti
 */
class ComplexMetadataTest {

    @Test
    fun test() {
        val m1 = SimpleMetadataTest.createSample("atomic", Subtype.CHAR, 10, 0, 1)
        val m2 = SimpleMetadataTest.createSample("array", Subtype.CHAR, 10, 1, 2)
        val children = ArrayList<Metadata>()
        children.add(m1)
        children.add(m2)
        val c1 = createSample("complex", Subtype.OBJ, 0, 0, 0, 1, Domain.EMPTY, EMPTY, EMPTY, children)

        assertEquals(m1, c1.children<Metadata>()[0])
        assertEquals(m2, c1.children<Metadata>()[1])
        //assertEquals(30, c1.serialMaxSize()); // m1 = 10 m2 = 20

        // Basic Methods
        val c2 = createSample("complex", Subtype.OBJ, 0, 0, 0, 1, Domain.EMPTY, EMPTY, EMPTY, children)

        assertTrue(c1 == c2)
        assertTrue(c1.hashCode() == c2.hashCode())
        assertTrue(c1.toString() == c2.toString())
        assertTrue(c2 == c1)
        assertTrue(c1 == c1)
        assertFalse(c1 == null)
        val o = "abc"
        assertFalse(c1.equals(o))
    }

    companion object {

        fun createSample(name: String, subtype: Subtype, size: Int, scale: Int, minOccurs: Int,
                         maxOccurs: Int, domain: Domain, defaultValue: String, tags: String, children: List<Metadata>?): ComplexMetadata {
            val metaInfo = MetaInfo(31, name, subtype, size, 0, minOccurs, maxOccurs, domain, defaultValue, tags)
            return ComplexMetadata(metaInfo, children!!)

        }

        internal fun createSample(name: String, subtype: Subtype, size: Int, minOccurs: Int, maxOccurs: Int): ComplexMetadata {
            val domain = domainOf("white", "black", "red")
            return createSample(name, subtype, size, 0, minOccurs, maxOccurs, domain, "red", "XYZ", null)

        }
    }

}
